service/kernel/data_management/train_test_split.cpp

```diff
@@ -0,0 +1,203 @@
+
+const size_t BLOCK_CONST = 10240;
+const size_t THREADING_BORDER = 12800000;
+
+template <typename Func>
+void runBlocks(const bool inParallel, const size_t nBlocks, Func func)
+{
+    if (inParallel)
+        daal::threader_for(nBlocks, nBlocks, [&](size_t i) { func(i); });
+    else
+        for (size_t i = 0; i < nBlocks; ++i) func(i);
+}
+
+template <typename DataType, typename IdxType>
+services::Status splitColumn(NumericTable & inputTable, NumericTable & trainTable, NumericTable & testTable,
+                             const IdxType * trainIdx, const IdxType * testIdx,
+                             const size_t nTrainRows, const size_t nTestRows, const size_t iCol)
+{
+    services::Status s;
+    BlockDescriptor<DataType> origBlock, trainBlock, testBlock;
+    inputTable.getBlockOfColumnValues(iCol, 0, nTrainRows + nTestRows, readOnly, origBlock);
+    trainTable.getBlockOfColumnValues(iCol, 0, nTrainRows, readWrite, trainBlock);
+    testTable.getBlockOfColumnValues(iCol, 0, nTestRows, readWrite, testBlock);
+    const DataType * origDataPtr = origBlock.getBlockPtr();
+    DataType * trainDataPtr = trainBlock.getBlockPtr();
+    DataType * testDataPtr = testBlock.getBlockPtr();
+    DAAL_CHECK_MALLOC(origDataPtr);
+    DAAL_CHECK_MALLOC(trainDataPtr);
+    DAAL_CHECK_MALLOC(testDataPtr);
+
+    const size_t BLOCK_SIZE = BLOCK_CONST;
+    const size_t nTrainBlocks = nTrainRows / BLOCK_SIZE ? nTrainRows / BLOCK_SIZE : 1;
+    const size_t nTestBlocks = nTestRows / BLOCK_SIZE ? nTestRows / BLOCK_SIZE : 1;
+    const int nTrainSurplus = nTrainRows / BLOCK_SIZE ? nTrainRows % BLOCK_SIZE : nTrainRows - BLOCK_SIZE;
+    const int nTestSurplus = nTestRows / BLOCK_SIZE ? nTestRows % BLOCK_SIZE : nTestRows - BLOCK_SIZE;
+
+    runBlocks(nTrainRows > THREADING_BORDER, nTrainBlocks, [&](size_t iBlock) {
+        const size_t start = iBlock * BLOCK_SIZE;
+        const size_t end   = iBlock == nTrainBlocks - 1 ? start + BLOCK_SIZE + nTrainSurplus : start + BLOCK_SIZE;
+
+        for (size_t i = start; i < end; ++i)
+            trainDataPtr[i] = origDataPtr[trainIdx[i]];
+    });
+
+    runBlocks(nTestRows > THREADING_BORDER, nTestBlocks, [&](size_t iBlock) {
+        const size_t start = iBlock * BLOCK_SIZE;
+        const size_t end   = iBlock == nTestBlocks - 1 ? start + BLOCK_SIZE + nTestSurplus : start + BLOCK_SIZE;
+
+        for (size_t i = start; i < end; ++i)
+            testDataPtr[i] = origDataPtr[testIdx[i]];
+    });
+
+    return s;
+}
+
+template <typename DataType, typename IdxType>
+services::Status splitRows(NumericTable & inputTable, NumericTable & trainTable, NumericTable & testTable,
+                           const IdxType * trainIdx, const IdxType * testIdx,
+                           const size_t nTrainRows, const size_t nTestRows, const size_t nColumns)
+{
+    services::Status s;
+    const size_t BLOCK_SIZE = BLOCK_CONST / nColumns ? BLOCK_CONST / nColumns : 1;
```

Comment:
```cpp
const size_t BLOCK_SIZE = max(1, BLOCK_CONST / nColumns);
```

Reply:
The same in other places

Reply:
Block borders and consts computation is corrected.